docs/arch/adr-011-kurl-addon-operator.md

```diff
@@ -0,0 +1,96 @@
+
+## Solution
+
+The operator will be implemented with the following objectives in mind:
+
+- It will be scaffolded using [kubebuilder](https://book.kubebuilder.io/introduction.html).
+- It will be developed using [SOLID](https://en.wikipedia.org/wiki/SOLID) principles and operator best practices.
+- It will distribution agnostic. I.e. it will not be tightly coupled to the kURL Kubernetes distribution.
+- It will manage application add-ons _only_ and _exclude_ managing CRI and CNI resources.
+- It will install and confiure the KOTS admin console.
+- It can be installed on an existing [kURL](https://kurl.sh/) cluster with only Kubernetes, containerd and flannel.
+- The installation mechanism for add-ons will be through the use of Helm Charts.
+- By default the add-on configuration will be opinionated by us but we will enable the end-user to override the configuration for maximum flexibility. Vendors and users will be able to override the configuration using a standard [Helm values](https://helm.sh/docs/chart_template_guide/values_files/) file.
+- It will rely as much as possible on Helm for managing the lifecyle of the add-ons.
+
+### The `Installer` API
+
+The `Installer` API defines a resource for installing specific add-ons.
+
+```yaml
+apiVersion: cluster.kurl.sh/v2beta1
+kind: Installer
+metadata:
+  name: installer-sample
+spec:
+  channel: "1.27"
+  minio:
+    enabled: true
+    configOverride:
+        persistence:
+            enabled: true
+        replicas: 1
+        mode: standalone
+
+```
+
+### The Controllers
+
+The operator will have at minimum one controller which ensures the `Installer` CR is reconciled with the cluster periodically. The `Installer` controller (main controller) will delegate responsiblity for managing add-ons to other controllers. The reconcilation for the main controller will consist of determining the add-ons to be managed from the `Installer` CR and creating a controller for each add-on (addon-on controller).
+
+The add-on controller will be responsible for
+- verifying the requirements for the add-on are met
+- managing the add-on itself
+- ensure add-on is in desired state
+- update the add-on CR status
+
+### Managing the Add-ons
+
+How the add-ons will be managed is still an open question. We have a couple of options:
+
+_Use the Helm Controller provided by Flux_
+
+The [Flux Helm Controller](https://fluxcd.io/flux/components/helm/helmreleases/) affords us the following benefits:
+- Stable [HelmRelease API](https://fluxcd.io/flux/components/helm/api/)
+- Helm release [dependencies](https://fluxcd.io/flux/components/helm/helmreleases/#helmrelease-dependencies)
+- Support for Helm install, upgrade, test, rollback and uninstall
```

Comment:
In our prototype implementation [here](https://github.com/replicatedhq/brainstorming-ekco/pull/2/files#diff-8e7acbce7dcceea6f611e0a7e681d52cfc12fa9f4678fee8cebfa1e3bdbe5ec2), we have demonstrated the ability to interact with Helm for installing, uninstalling, upgrading, retrieving release information, and converting the release status for a HelmChart. This functionality is achieved with pure Helm, utilizing a single file with just a few lines of code. Since the logic is still handled by the Helm dependency, the implementation remains concise.

Considering this, I initially anticipated that Flux would provide similar capabilities for performing these operations. However, I don't necessarily view such features in Flux as a pro for us. I think we need to looking for what Flux provide and offer for us that would not be easily achieved by only calling Helm. 

However, if we would not go in the direction to build an Operator and we would only tooling the solution then I think Flux would be a game change because would provide all those things. But that also would mean we would not able to do any extra checks or operations before and after install / uninstall / update a Helm solution.

Reply:
anything valuable we find in the helmrelease_controller.go i assume we would have to build ourselves

https://github.com/fluxcd/helm-controller/blob/main/internal/controllers/helmrelease_controller.go

Reply:
We'll need a controller in any scenario to manage statuses, finalizers, and so on. Flux, while a robust solution, is generic and therefore doesn't address our specific needs and requirements.

Our objective is highly targeted, and it's important to remember that we can't attempt to update or delete any resource created and managed by Flux. Doing so would trigger the control-loop which could lead to unintended consequences. In many scenarios, this could result in Flux re-executing all operations to ensure its processes are maintained, as that's what a control-loop is designed to do. Consequently, the issue we tried to resolve might resurface. Alternatively, we could encounter race or concurrency conditions, as both we and Flux would be trying to update/change the same resources simultaneously.

Flux isn't tailored to cater to specific needs, like what we're seeking here, nor is it designed for the kind of integration and usage we're proposing. Its main function is to provide a tool capable of automating the process of installing, updating, or uninstalling a HelmChart, without considering any additional requirements. For example, it wouldn't accommodate a scenario where Kots is configured with N, implying that X must be present. In situations where you try to uninstall X without Y, it won't work. Flux doesn't account for these kinds of complexities.

Reply:
If we use flux, I suggest we never modify the Kubernetes resources themselves and instead modify values through the HelmChart CR. If we ever had to we can use the suspend property of the [CRD](https://fluxcd.io/flux/components/helm/helmreleases/#specification).

Flux is a reconciliation loop. If one chart has a precondition it will fail until that precondition is met and then succeed.

I feel the beauty in the flux helm controller is it is built to do one thing only and that is manage the lifecycle of helm charts.

Reply:
> anything valuable we find in the helmrelease_controller.go i assume we would have to build ourselves
>
>https://github.com/fluxcd/helm-controller/blob/main/internal/controllers/helmrelease_controller.go

I think this is a really insightful comment. Has anyone reviewed the controller? I'm curious how much of this controller do we think we need to implement for a robust solution?

Reply:
@emosbaugh,

> If we use flux, I suggest we never modify the Kubernetes resources themselves and instead modify values through the HelmChart CR. If we ever had to we can use the suspend property of the [CRD](https://fluxcd.io/flux/components/helm/helmreleases/#specification)

If we choose to adopt Flux, it would imply that we could no longer directly install, delete, or update HelmCharts or Custom Resource Definitions (CRDs) in our controllers. Additionally, we would not be able to manage any resources created and managed by Flux during the installation of these HelmCharts.

However, it's important to note that our business requirements necessitate certain actions. These include ensuring seamless migrations, implementing workarounds for specific issues, and automating steps that are presently manual and only applicable within our particular operational context to guarantee smooth upgrades. Consequently, it's crucial that we retain control over these processes. While Flux is a powerful tool, it doesn't cater to the kind of specific use-cases that we often need to address. Therefore, we need a solution that accommodates our unique needs while providing comprehensive control over our resources. 

If we try to manipulated the same resources then you are proposing a solution that violates the SOLID principle and suggests the use of two overlapping Operators, it would introduce potential challenges related to race conditions and concurrency. It's important to note that having two instances of the same Operator is not feasible due to these reasons, and the idea of having two separate Operator solutions responsible for the same tasks clearly goes against the Single Responsibility Principle, as per SOLID principles

Reply:
Hi @chris-sanders, 

I've checked a little the[ helmrelease_controller.go](https://github.com/fluxcd/helm-controller/blob/main/internal/controllers/helmrelease_controller.go) file from FluxCD's Helm Controller. Then, for me although it contains a wide range of features, not all shows that are relevant to our specific needs, and crucially, it doesn't address the unique requirements that we have. As a result, our implementation to meet our business requirements—striving for full automation to minimize human intervention, and utilizing a reduced, opinionated specification—**would IHMO not mirror Flux's approach.**

If the concern is about the potential effort involved in developing a similar solution, could we not consider Flux's implementation as a reference point? If our paths do align—**which I currently do not foresee see too much beyond mainly installing/uninstalling/deleting HelmCharts—Flux's** code could serve as a useful guide to expedite our progress. WDTY?

Following a couple of points. 

## Flux reconciliation logic is not the same approach proposed in the prototype

The approach of our prototype using Helm for reconciliation significantly differs from Flux's diff-based approach (see [here](https://github.com/fluxcd/helm-controller/blob/main/internal/controllers/helmrelease_controller.go#L162-L170)). Instead of relying on diffing the current cluster state against the desired state, our prototype takes a different approach as proposed [here](https://github.com/replicatedhq/brainstorming-ekco/pull/2/files#diff-98c935b2ef1adddcc8355cdf8f61bf1adb3c14ab90de326a27874d5bdbd7f6dbR130-R256).

In our prototype, we proposed directly handling reconciliation based on the desired state specified through user input into the Installer/Kurl CR, our defaults, and specific HelmChart versions. Unlike Flux's Diff() function, which shows to calculates the changes needed to reconcile the current state of Kubernetes resources with the state defined in the Helm release, our custom Operator avoids this diffing process. 

The diff approach shows that can bring for example:

- False Positives: Diffing relies on comparing the current cluster state with the desired state defined in configuration files. However, this comparison may result in false positives, indicating differences that are not actual drift. This can lead to unnecessary and potentially disruptive corrective actions.
- Continuous Drift: Depending solely on diff-based drift detection and correction may result in a continuous cycle of corrections, causing unnecessary system churn and instability. 
- Overwriting Manual Interventions and Customizations: The diff approach shows that cannot consider manual changes or configurations made to the cluster, leading to unintended consequences when the system attempts to automatically correct these changes. This can result in disruptions or loss of customized configurations that could not be addressed via the chart values for example.  

### Managing Dependencies and Pre-requirements/Conditions

In our custom solution using pure Helm, we prioritize validating that all pre-requisites for installing, uninstalling, or deleting an AddOn are satisfied for each specific scenario. We don't need to rely on a generic mode, which might not comprehensively cover our needs. For instance, while handling dependencies, Flux merely checks if a HelmRelease is installed and declares success (see [here](https://github.com/fluxcd/helm-controller/blob/main/internal/controllers/helmrelease_controller.go#L258-L272)).

When we were doing our prototype and demo, we checked that there are circumstances where shows that a HelmRelease might report that a solution is deployed when, in fact, it isn't fully functional or installed properly. In our controller, we wouldn't manage things this way. Instead, we can prioritize validating if ANY specific requirements in line with our business rules and customer demands - are not just deployed, but also HEALTHY and aligned with our particular opniated spec configurations, before performing any next step operation. This degree of detailed health-check is beyond Flux's capabilities for any pre-condition addon or dependency. As a result, our custom solution could provides superior assurance and control and grow according with the problems faced by the users.Therefore the custom approach has the potential to:

- mitigate scenarios where an attempted solution installation results in an unhealthy cluster state, often necessitating manual intervention—a common issue with the current KURL and a pain point for our users. 
- moreover, we would be able to effectively monitor and report status conditions, indicating when pre-requisite conditions are not met, as proposed [here](https://github.com/replicatedhq/brainstorming-ekco/issues/6). Thus, our approach using pure Helm enhances stability, user experience, and transparency of system status which is not achievable by adopting Flux. 
- prevent scenarios where an attempted solution installation could lead to an unhealthy cluster state --common user feedback and request. This minimizes the need for manual intervention, a significant issue with the current KURL and a recurrent user concern.

## Tracking the current state of the solutions

As previously emphasized, Flux is a generic solution and, consequently, lacks the capacity to comprehensively verify whether each specific Addon installation or upgrade has been successfully completed and is functioning healthily, particularly considering the unique intricacies of each specific solution.

In contrast, by retaining full control over our operations, we could perform granular checks on pre-requisites for each specific Addon, ensuring their individual compatibility and readiness within our specific system. This ties into the approach proposed [here](https://github.com/replicatedhq/brainstorming-ekco/issues/6). That can enables us to gain granular visibility into system health, pinpoint any issues, identify their causes, and rectify them effectively. 

Moreover, as demonstrated in our [prototype](https://github.com/replicatedhq/brainstorming-ekco/pull/2/files#diff-8e7acbce7dcceea6f611e0a7e681d52cfc12fa9f4678fee8cebfa1e3bdbe5ec2R132-R159), using pure Helm enables us to retrieve the status from Helm and convert it into our custom Status conditional structure with just a few lines of code.

### Remove the complexities to deal and parse Flux logs

In contrast, Flux would require us as discussed yesterday and shared for those that have been working with in many scenarios parse LOGS and not only monitor its CRS statuses, introducing a new layer of complexity. In POV, it would necessitate significantly more code implementation and maintenance. 

## Let's now speak about FLUX finalizers implementation

The Flux Controller handles the finalizer, a method that prevents deletion of a Custom Resource (CR) representing the solution on the cluster or the corresponding namespace where the resource is installed. It operates by simply verifying whether Helm has reported that the Chart was uninstalled, as evidenced [here](https://github.com/fluxcd/helm-controller/blob/74bc32e840e1b09e48036cb5553c62a71a49baba/internal/controllers/helmrelease_controller.go#L669-L725).

On the other hand, when using our custom Operator with pure Helm, we have the flexibility to perform additional operations prior to deletion. This allows us to prevent unintended consequences such as data loss, which is discussed in greater detail [here](https://github.com/replicatedhq/kURL/pull/4488#discussion_r1197553410). The use of Helm gives us more granular control over the lifecycle of our resources, a level of detail not provided by Flux's approach.